coresimd/src/runtime/linux/auxv.rs

```diff
@@ -0,0 +1,140 @@
+//!
+//! The keys used in the aux vector are platform dependent. For Linux, they are
+//! defined in [linux/auxvec.h][auxvec_h]. The hardware capabilities of a given
+//! CPU can be queried with the  `AT_HWCAP` and `AT_HWCAP2` keys.
+//!
+//! There is no perfect way of reading the auxiliary vector.
+//!
+//! - `coresimd`: if `getauxval` is available, `coresimd` will try to use it.
+//! - `stdsimd`: if `getauxval` is not available, it will try to read
+//! `/proc/self/auxv`, and if that fails it will try to read `/proc/cpuinfo`.
+//!
+//! For more information about when `getauxval` is available check the great
+//! [`auxv` crate documentation][auxv_docs].
+//!
+//! [auxvec_h]: https://github.com/torvalds/linux/blob/master/include/uapi/linux/auxvec.h
+//! [auxv_docs]: https://docs.rs/auxv/0.3.3/auxv/
+
+/// Key to access the CPU Hardware capabilities bitfield.
+pub const AT_HWCAP: usize = 16;
+/// Key to access the CPU Hardware capabilities 2 bitfield.
+pub const AT_HWCAP2: usize = 26;
+
+/// Cache HWCAP bitfields of the ELF Auxiliary Vector.
+///
+/// If an entry cannot be read all the bits in the bitfield
+/// are set to zero.
+#[cfg(any(target_arch = "arm", target_arch = "powerpc64"))]
+#[derive(Debug, Copy, Clone)]
+pub struct AuxVec {
+    pub hwcap: usize,
+    pub hwcap2: usize,
+}
+
+/// Cache HWCAP bitfields of the ELF Auxiliary Vector.
+///
+/// If an entry cannot be read all the bits in the bitfield
+/// are set to zero.
+#[cfg(target_arch = "aarch64")]
+#[derive(Debug, Copy, Clone)]
+pub struct AuxVec {
+    pub hwcap: usize,
+}
+
+pub mod libc {
+    use super::*;
+
+    // Since Rust does not support weak lining yet, we use the
+    // system C compiler to weakly-link `getauxval`.
+    extern "C" {
+        fn getauxval_wrapper(key: usize, success: *mut usize) -> i32;
+    }
+
+    /// Returns the value of the ELF Auxiliary Vector associated with `key`.
+    ///
+    /// This can fail if the auxiliary vector does not contain the key,
+    /// the final binary is not linked against a libc library containing
+    /// `getauxval`, etc.
+    ///
+    /// This only returns the value if reading the key from the auxiliary
+    /// vector properly succeeds.
+    fn getauxval(key: usize) -> Result<usize, ()> {
+        let mut auxv = 0;
+        unsafe {
+            match getauxval_wrapper(key, &mut auxv) {
+                1 => Ok(auxv),
+                _ => Err(()),
+            }
+        }
+    }
+
+    /// Computes the entries of the Auxiliary Vector cache by
+    /// calling libc's `getauxval(3)`.
+    pub fn auxv() -> Result<AuxVec, ()> {
+        if let Ok(hwcap) = getauxval(AT_HWCAP) {
+            #[cfg(target_arch = "aarch64")]
+            {
+                return Ok(AuxVec { hwcap });
+            }
+            #[cfg(any(target_arch = "arm", target_arch = "powerpc64"))]
+            {
+                if let Ok(hwcap2) = getauxval(AT_HWCAP2) {
+                    return Ok(AuxVec { hwcap, hwcap2 });
```

Comment:
This is a bit different from what @lu-zero . Now, either reading both fields succeeds, or the whole read fails. This allows that, in case reading one of them succeeds but reading the other fails, we can still try to read `/proc/self/auxv` or `/proc/cpuinfo` to obtain the rest of the information.

Basically either `getauxval` works 100% reliably, or we fall back to doing something slower but reliable.